facilitator/src/idl.rs

```diff
@@ -0,0 +1,561 @@
+            "type": "string",
+            "doc": "a name for this specific aggregation"
+        },
+        {
+            "name": "bins",
+            "type": "int",
+            "doc": "number of bins for this aggregation"
+        },
+        {
+            "name": "epsilon",
+            "type": "double",
+            "doc": "differential privacy parameter for local randomization before aggregation."
+        },
+        {
+            "name": "prime",
+            "type": "long",
+            "default": 4293918721,
+            "doc": "the value of prime p used in aggregation."
+        },
+        {
+            "name": "number_of_servers",
+            "type": "int",
+            "default": 2,
+            "doc": "the number of servers that will be involved in the aggregation."
+        },
+        {
+            "name": "hamming_weight",
+            "type": ["int", "null"],
+            "doc":  "If specified, the hamming weight of the vector will be verified during the validity check on the server."
+        },
+        {
+            "name": "batch_start_time",
+            "type": "long",
+            "logicalType": "timestamp-millis",
+            "doc": "time range information for the shares in this batch."
+        },
+        {
+            "name": "batch_end_time",
+            "type": "long",
+            "logicalType": "timestamp-millis",
+            "doc": "time range information for the shares in this batch."
+        }
+    ]
+}
+"#;
+
+#[derive(Debug, Deserialize, Serialize, PartialEq)]
+pub struct IngestionHeader {
+    pub batch_uuid: Uuid,
+    pub name: String,
+    pub bins: i32,
+    pub epsilon: f64,
+    pub prime: i64,
+    pub number_of_servers: i32,
+    pub hamming_weight: Option<i32>,
+    pub batch_start_time: i64,
+    pub batch_end_time: i64,
+}
+
+impl IngestionHeader {
+    pub fn read_header(reader: impl Read) -> Result<IngestionHeader, Error> {
+        let schema = Schema::parse_str(INGESTION_HEADER_SCHEMA).map_err(|e| Error::AvroError(e))?;
+        let mut reader = Reader::with_schema(&schema, reader).map_err(|e| Error::AvroError(e))?;
+
+        // We expect exactly one record in the reader and for it to be an ingestion header
+        let header = match reader.next() {
+            Some(Ok(h)) => h,
+            Some(Err(e)) => return Err(Error::AvroError(e)),
+            None => {
+                return Err(Error::MalformedHeaderError(
+                    "no records found in reader".to_owned(),
+                ));
+            }
+        };
+        if let Some(_) = reader.next() {
+            return Err(Error::MalformedHeaderError(
+                "excess header in reader".to_owned(),
+            ));
+        }
+
+        from_value::<IngestionHeader>(&header).map_err(|e| Error::AvroError(e))
+    }
+
+    pub fn write_header<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
+        let schema = Schema::parse_str(INGESTION_HEADER_SCHEMA).map_err(|e| Error::AvroError(e))?;
```

Comment:
I agree! The reason I didn't do that is because then I couldn't see how I would go about providing an error string to the `facilitator` level error that we wrap underlying errors with.

Reply:
Yeah, it made a bit more sense afterwards when you added the message strings. Back in this commit you didn't have any message yet and so the errors could be auto-converted.

One interesting thing you can do re. this though is to actually make the message optional in the error, so you'd do something like this:

```rust
#[derive(Debug)]
pub enum ErrorKind {
    AvroError(avro_rs::Error),
    MalformedHeaderError,
    MalformedSignatureError,
    MalformedDataPacketError,
    // ...
}

#[derive(Debug)]
pub struct Error {
    pub detail: String,
    pub kind: ErrorKind,
}
```

such that you decouple the message from the individual errors, and then to more conveniently manipulate instances, add some helper methods:

```rust
impl Error {
    pub fn new(kind: ErrorKind) -> Error {
        Error { detail: String::new(), kind }
    }

    // Taking `Into<String>` allows you to take either a `&str` or an actual `String` as an argument
    pub fn with_message(self, detail: impl Into<String> ) -> Error {
        Error { detail: detail.into(), ..self }
    }
}
```

Now you can just add `From` impls for various library error types:

```rust
impl From<avro_rs::Error> for Error {
    fn from(e: avro_rs::Error) -> Self {
        Error::new(AvroError(e))
        // If you have a "default" message, such as if AvroError itself has enough information to
        // come up with some message, you could call `.with_message` here to pre-populate one.
    }
}

// etc.
```

And the usage becomes:

```rust
fn f() -> Result(Foo, Error) {
    let x = my_avro_call_that_can_fail()?;
    // do stuff
    Ok(foo)
}
```

Which is nice! `?` will automatically use the `From` impl to convert the error in the `Result`. But if we want to add a message it's a bit more annoying, since you need to manually convert to our error first:

```rust
let x = my_avro_call_that_can_fail().map_err(|e| Error::from(e).with_message("something went wrong"))?;
```

Which is ok, but still kind of un-ergonomic. More impls to the rescue! First we create a helper trait. This allows us to attach new methods to existing types:

```rust
pub trait ResultWithMessage<T> {
    fn with_message(self, detail: impl Into<String>) -> Result<T, Error>;
}
```

Now, we can implement it for any `Result` with an error that we know how to convert:

```rust
impl<T, E> ResultWithMessage<T> for Result<T, E>
where
    E: Into<Error>,
{
    fn with_message(self, detail: impl Into<String>) -> Result<T, Error> {
        self.map_err(|e| e.into().with_message(detail))
    }
}
```

(Note that this includes `Result<T, Error>` itself, since there's a blanket implementation of `impl From<T> for T` which is a no-op, so we don't need to add a separate impl for that.)

And now we can just annotate results as we see fit:

```rust
use error::ResultWithMessage; // needed if the trait is defined in another module

let x = my_avro_call_that_can_fail().with_message("something went wrong")?;
```

Reply:
Of course there are libraries that have this kind of framework already in place so we should probably just use one of those. :)